documentation/sections/let_const.md

```diff
@@ -0,0 +1,103 @@
+(function() {   // We generate the entire file within a closure by default, to avoid variabe leakage.
+  var a, b;     // a, b are visible in all remaining scopes as a result of closures.
+                // however they are NOT visible globally, only within this file / module.
+  a = 5;
+
+  b = function() {
+    var c;        // c is not seen outside of this scope.
+    return c = 6;
+  };
+
+}).call(this);
+
+```
+
+Aha!  Right there in that function, you could accidentally override `a`!  
+
+Well this is true. But consider [this conversation](https://github.com/jashkenas/coffeescript/issues/238#issuecomment-153502) in the issues:
+
+> It's certainly possible to accidentally blow away an outer variable within an inner scope, but (in Ruby and Python), this turns out not to be a very common problem. If you don't nest functions too deeply, don't introduce a lot of globals, and give things proper names, it never arises.
+
+And here is another useful comment by jashkenas
+
+> And if they do clash, shadowing the variable is the wrong answer. It completely prevents you from making use of the original value for the remainder of the current scope. Shadowing doesn't fit well in languages with closures-by-default ... if you've closed over that variable, then you should always be able to refer to it.
+
+If you take the time to name things well, you actually have to go out of your way to clobber another variable. The trade off for making you take more time naming your variables is one less keyword to have to type over and over.
+
+Hrmmm. Good point. But what about `const`?!? I REALLY want to make everything constant!
+
+### The `const` Keyword
+
+`const` is another typical keyword that doesn't really add significant value. Primarily `const` is used to avoid accidental reassignments. In practice this is just like `let` above, make sure you are consistant with scope, and name your variables well. 
+
+`const` also causes a dilemma for CoffeeScript, because suddenly we can get block level shadowing, but it's only const!  Well then why not include `let`!  Well hopefully you see why we haven't included `let` by now. 
+So how the heck do I make things constant? 
+
+```coffeescript
+
+a = 0
+b = 1
+` const c = a + b `
+
+c++   # Error - c is a const.
+
+```
+
+
+In reality the const keyword refers to the variable itself, _not always the value_ of that object. 
+
+```javascript
+  
+const x = 5;
+const y = {};
+x++;          // error.  Assignment to const
+y[x] = x;     // {'5': 4}  Wait what!??
+y = x;        // error.  Assignment to const
+
+```
+
+In other words, the statement const x++ doesn't let you change the variable as you would expect. However it WILL let you change the contents of y, because you aren't changing what y is pointing at. It is still the same object! 
+
+You actually have to 'freeze' an object to make it a constant. Use [Object.freeze(obj)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze).
```

Comment:
This section is way too long. I would try to reduce it to at most three paragraphs, and link to other sources to add more detail. Also try to write it in a dry tone like the rest of the documentation. Something like:

When CoffeeScript was designed, `var` was intentionally omitted. This was to spare developers the mental housekeeping of needing to worry about variable _declaration_ (`var foo`) as opposed to variable _assignment_ (`foo = 1`). The CoffeeScript compiler automatically takes care of declaration for you, by generating `var` statements at the top of every function scope. This makes it impossible to accidentally declare a global variable.

`let` and `const` add a useful ability to JavaScript in that you can use them to declare variables within a _block_ scope, for example within an `if` statement body or a `for` loop body, whereas `var` always declares variables in the scope of an entire function. When CoffeeScript 2 was designed, there was much discussion of whether this functionality was useful enough to outweigh the simplicity offered by never needing to consider variable declaration in CoffeeScript. In the end, it was decided that the simplicity was valued more. Ultimately there is nothing provided by `let` or `const` that you cannot achieve with `var` and careful naming of variables, avoiding reusing variable names within the same function scope.